rust/lance-index/src/scalar/inverted.rs

```diff
@@ -262,86 +276,89 @@ impl TokenSet {
     }
 
     async fn load(reader: Arc<dyn IndexReader>) -> Result<Self> {
-        let mut tokens = Vec::new();
-        let mut ids = Vec::new();
-        let mut frequencies = Vec::new();
+        let mut tokens = HashMap::new();
+        let mut next_id = 0;
         for i in 0..reader.num_batches().await {
             let batch = reader.read_record_batch(i).await?;
-            let token_col = batch[TOKEN_COL].as_string::<i32>();
+            let token_col = batch[TOKEN_COL].as_string::<i64>();
             let token_id_col = batch[TOKEN_ID_COL].as_primitive::<datatypes::UInt32Type>();
             let frequency_col = batch[FREQUENCY_COL].as_primitive::<datatypes::UInt64Type>();
 
-            tokens.extend(token_col.iter().map(|v| v.unwrap().to_owned()));
-            ids.extend(token_id_col.iter().map(|v| v.unwrap()));
-            frequencies.extend(frequency_col.iter().map(|v| v.unwrap()));
+            for ((token, token_id), frequency) in token_col
+                .iter()
+                .zip(token_id_col.iter())
+                .zip(frequency_col.iter())
+            {
+                let token = token.unwrap();
+                let token_id = token_id.unwrap();
+                let frequency = frequency.unwrap();
```

Comment:
You can avoid the null checks / unwraps by iterator over the values buffer:

```suggestion
                .zip(token_id_col.values().iter())
                .zip(frequency_col.values().iter())
            {
                let token = token.unwrap();
```